pkg/controller/registry/resolver/evolver.go

```diff
@@ -153,7 +162,65 @@ func (e *NamespaceGenerationEvolver) queryForRequiredAPIs() error {
 	return nil
 }
 
+func (e *NamespaceGenerationEvolver) downgradeUpdates() error {
+	// no need to attempt downgrades
+	if len(e.gen.MissingAPIs()) == 0 {
```

Comment:
Super nit: the caller for this function might be more obvious if this check happened outside the method itself.

Reply:
this is just keeping in line with the rest of the object, which has a series of methods which may or may not affect the generation - adding the check in `Evolve` would make it read weirdly, IMO